File: handwritten_ocr.java
// handwritten_ocr.java — Java версия

import net.sourceforge.tess4j.Tesseract;
import net.sourceforge.tess4j.TesseractException;
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class handwritten_ocr {
    public static void main(String[] args) throws Exception {
        String imagePath = null;
        String lang = "eng";
        boolean batch = false;
        String output = null;

        for (int i = 0; i < args.length; i++) {
            if (args[i].equals("--lang") || args[i].equals("-l")) {
                lang = args[++i];
            } else if (args[i].equals("--batch") || args[i].equals("-b")) {
                batch = true;
            } else if (args[i].equals("--output") || args[i].equals("-o")) {
                output = args[++i];
            } else if (imagePath == null) {
                imagePath = args[i];
            }
        }

        if (imagePath == null) {
            System.out.println("Usage: java handwritten_ocr <image> [--lang eng] [--batch]");
            System.exit(1);
        }

        System.out.println("\u001B[36m✍️ OCR Handwritten Reader (Java)\u001B[0m");

        Tesseract tesseract = new Tesseract();
        tesseract.setLanguage(lang);
        tesseract.setPageSegMode(6); // Single block of text
        tesseract.setOcrEngineMode(3); // LSTM

        if (batch || Files.isDirectory(Paths.get(imagePath))) {
            processBatch(tesseract, imagePath);
        } else {
            processSingle(tesseract, imagePath, output);
        }
    }

    private static void processSingle(Tesseract tesseract, String imagePath, String output) throws TesseractException {
        long start = System.currentTimeMillis();
        String text = tesseract.doOCR(new File(imagePath));
        double elapsed = (System.currentTimeMillis() - start) / 1000.0;

        printResult(text, elapsed);

        if (output == null) {
            output = imagePath.replaceFirst("\\.[^.]+$", "") + ".txt";
        }
        try (FileWriter fw = new FileWriter(output)) {
            fw.write(text);
        } catch (IOException e) {
            System.out.println("Ошибка сохранения.");
        }
        System.out.println("💾 Сохранено: " + output);
    }

    private static void processBatch(Tesseract tesseract, String dir) throws TesseractException {
        String[] exts = {".jpg", ".jpeg", ".png", ".bmp", ".tiff"};
        File folder = new File(dir);
        List<File> files = new ArrayList<>();
        for (File f : folder.listFiles()) {
            if (f.isFile()) {
                String ext = f.getName().toLowerCase();
                for (String e : exts) {
                    if (ext.endsWith(e)) files.add(f);
                }
            }
        }

        if (files.isEmpty()) {
            System.out.println("❌ Нет поддерживаемых изображений.");
            return;
        }

        System.out.println("📁 Найдено " + files.size() + " изображений.");
        for (int i = 0; i < files.size(); i++) {
            System.out.printf("\n[%d/%d] Обработка: %s\n", i+1, files.size(), files.get(i).getName());
            String text = tesseract.doOCR(files.get(i));
            String outFile = files.get(i).getName().replaceFirst("\\.[^.]+$", "") + ".txt";
            try (FileWriter fw = new FileWriter(outFile)) {
                fw.write(text);
            } catch (IOException e) {}
            System.out.println("💾 Сохранено: " + outFile);
        }
    }

    private static void printResult(String text, double elapsed) {
        System.out.println("\n\u001B[32mРезультат:\u001B[0m");
        System.out.println("─".repeat(50));
        System.out.println(text);
        System.out.println("─".repeat(50));

        String[] words = text.split("\\s+");
        int wordCount = text.isEmpty() ? 0 : words.length;
        int charCount = text.length();
        System.out.println("\u001B[36m📊 Статистика:\u001B[0m");
        System.out.println("  Слов: " + wordCount);
        System.out.println("  Символов: " + charCount);
        System.out.println("  Время: " + String.format("%.2f", elapsed) + " сек");
    }
}

File: handwritten_ocr.cs
// handwritten_ocr.cs — C# версия

using System;
using System.IO;
using System.Text.Json;
using System.Collections.Generic;
using System.Linq;
using Tesseract;

class HandwrittenOCR {
    static void Main(string[] args) {
        string imagePath = null;
        string lang = "eng";
        bool batch = false;
        string output = null;

        for (int i = 0; i < args.Length; i++) {
            if (args[i] == "--lang" || args[i] == "-l") {
                lang = args[++i];
            } else if (args[i] == "--batch" || args[i] == "-b") {
                batch = true;
            } else if (args[i] == "--output" || args[i] == "-o") {
                output = args[++i];
            } else if (imagePath == null) {
                imagePath = args[i];
            }
        }

        if (imagePath == null) {
            Console.WriteLine("Usage: dotnet run <image> [--lang eng] [--batch]");
            return;
        }

        Console.WriteLine("\u001B[36m✍️ OCR Handwritten Reader (C#)\u001B[0m");

        using (var engine = new TesseractEngine("./tessdata", lang, EngineMode.Default)) {
            engine.DefaultPageSegMode = PageSegMode.SingleBlock;
            if (batch || Directory.Exists(imagePath)) {
                ProcessBatch(engine, imagePath);
            } else {
                ProcessSingle(engine, imagePath, output);
            }
        }
    }

    static void ProcessSingle(TesseractEngine engine, string imagePath, string output) {
        var start = DateTime.Now;
        using (var img = Pix.LoadFromFile(imagePath)) {
            using (var page = engine.Process(img)) {
                string text = page.GetText().Trim();
                var elapsed = (DateTime.Now - start).TotalSeconds;

                PrintResult(text, elapsed);

                if (output == null) {
                    output = Path.GetFileNameWithoutExtension(imagePath) + ".txt";
                }
                File.WriteAllText(output, text);
                Console.WriteLine($"💾 Сохранено: {output}");
            }
        }
    }

    static void ProcessBatch(TesseractEngine engine, string dir) {
        var exts = new[] { ".jpg", ".jpeg", ".png", ".bmp", ".tiff" };
        var files = Directory.GetFiles(dir).Where(f => exts.Contains(Path.GetExtension(f).ToLower())).ToList();

        if (files.Count == 0) {
            Console.WriteLine("❌ Нет поддерживаемых изображений.");
            return;
        }

        Console.WriteLine($"📁 Найдено {files.Count} изображений.");
        var results = new List<Dictionary<string, string>>();

        for (int i = 0; i < files.Count; i++) {
            Console.WriteLine($"\n[{i+1}/{files.Count}] Обработка: {Path.GetFileName(files[i])}");
            using (var img = Pix.LoadFromFile(files[i])) {
                using (var page = engine.Process(img)) {
                    string text = page.GetText().Trim();
                    string outFile = Path.GetFileNameWithoutExtension(files[i]) + ".txt";
                    File.WriteAllText(outFile, text);
                    Console.WriteLine($"💾 Сохранено: {outFile}");
                    results.Add(new Dictionary<string, string> { ["file"] = Path.GetFileName(files[i]), ["text"] = text });
                }
            }
        }

        string json = JsonSerializer.Serialize(results, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText("ocr_results.json", json);
        Console.WriteLine($"\n💾 Сохранено JSON: ocr_results.json");
    }

    static void PrintResult(string text, double elapsed) {
        Console.WriteLine($"\n\u001B[32mРезультат:\u001B[0m");
        Console.WriteLine(new string('─', 50));
        Console.WriteLine(text);
        Console.WriteLine(new string('─', 50));

        var words = text.Split(new[] { ' ', '\n' }, StringSplitOptions.RemoveEmptyEntries).Length;
        Console.WriteLine($"\u001B[36m📊 Статистика:\u001B[0m");
        Console.WriteLine($"  Слов: {words}");
        Console.WriteLine($"  Символов: {text.Length}");
        Console.WriteLine($"  Время: {elapsed:F2} сек");
    }
}
